File: connectivity/icepolicy/endpoints.go
// Package icepolicy owns trusted, transport-independent ICE selection policy.
// Construct pools only from build, deployment, or explicit local configuration.
package icepolicy

import (
	"encoding/json"
	"errors"
	"fmt"
	"hash/fnv"
	"net"
	"sort"
	"strconv"
	"strings"
	"time"
)

const (
	EndpointsPerProfile     = 2
	EndpointsPerWave        = 4
	MaximumCatalogEndpoints = 512
	EndpointFailureCooldown = 30 * time.Second
)

var ErrConfiguration = errors.New("invalid trusted ICE endpoint configuration")

type Endpoint struct {
	ID            string `json:"id"`
	URL           string `json:"url"`
	Region        string `json:"region"`
	FailureDomain string `json:"failureDomain"`
	Provider      string `json:"provider"`
	Family        string `json:"family"`
	Trust         string `json:"trust"`
	Priority      int    `json:"priority"`
	Enabled       bool   `json:"enabled"`
}

// ICEEndpointPool keeps catalog entries separate from enabled trusted endpoints.
type ICEEndpointPool struct{ endpoints []Endpoint }

func NewICEEndpointPool(entries []Endpoint) (ICEEndpointPool, error) {
	if len(entries) > MaximumCatalogEndpoints {
		return ICEEndpointPool{}, ErrConfiguration
	}
	seen := make(map[string]bool)
	for _, endpoint := range entries {
		if endpoint.ID == "" || seen[endpoint.ID] || !validEndpointURL(endpoint.URL) ||
			(endpoint.Family != "ipv4" && endpoint.Family != "ipv6" && endpoint.Family != "any") ||
			(endpoint.Trust != "reviewed" && endpoint.Trust != "local" && endpoint.Trust != "unreviewed") ||
			(endpoint.Enabled && endpoint.Trust == "unreviewed") || endpoint.Priority < 0 {
			return ICEEndpointPool{}, ErrConfiguration
		}
		seen[endpoint.ID] = true
	}
	return ICEEndpointPool{endpoints: append([]Endpoint(nil), entries...)}, nil
}

// ParseLocalConfiguration never resolves or contacts an endpoint.
func ParseLocalConfiguration(data []byte) (ICEEndpointPool, error) {
	var entries []Endpoint
	if err := json.Unmarshal(data, &entries); err != nil {
		return ICEEndpointPool{}, fmt.Errorf("%w: %w", ErrConfiguration, err)
	}
	return NewICEEndpointPool(entries)
}

func validEndpointURL(raw string) bool {
	if !strings.HasPrefix(raw, "stun:") || strings.ContainsAny(raw, "/?#@ \t\r\n") {
		return false
	}
	host, port, err := net.SplitHostPort(strings.TrimPrefix(raw, "stun:"))
	if err != nil || host == "" {
		return false
	}
	number, err := strconv.Atoi(port)
	return err == nil && number > 0 && number <= 65535
}

type SelectionRequest struct {
	NetworkGenerationID string
	WaveID              string
	Sequence            uint64
	Now                 time.Time
	UsedEndpointIDs     []string
	UsedFailureDomains  []string
	IPv4                bool
	IPv6                bool
}

type EndpointFact struct {
	EndpointID  string
	FailedUntil time.Time
}
type ProfileFact struct {
	ProfileID               string
	ServerReflexiveProduced bool
	FirstCandidateDelay     time.Duration
}
type GenerationFacts struct {
	NetworkGenerationID string
	Endpoints           []EndpointFact
	Profiles            []ProfileFact
}

// AttemptICEProfile is a snapshot: accessors copy slices so a live attempt cannot
// change configuration when the owner's pool or observations change.
type AttemptICEProfile struct {
	id, generation string
	endpoints      []Endpoint
}

func (p AttemptICEProfile) ID() string                  { return p.id }
func (p AttemptICEProfile) NetworkGenerationID() string { return p.generation }
func (p AttemptICEProfile) EndpointIDs() []string {
	out := make([]string, len(p.endpoints))
	for i, e := range p.endpoints {
		out[i] = e.ID
	}
	return out
}
func (p AttemptICEProfile) URLs() []string {
	out := make([]string, len(p.endpoints))
	for i, e := range p.endpoints {
		out[i] = e.URL
	}
	return out
}
func (p AttemptICEProfile) FailureDomains() []string {
	out := make([]string, len(p.endpoints))
	for i, e := range p.endpoints {
		out[i] = e.FailureDomain
	}
	return out
}

// RebindAttemptProfile rechecks an existing wave's membership against a new
// network generation without granting another endpoint exploration allowance.
func RebindAttemptProfile(profile AttemptICEProfile, request SelectionRequest, facts GenerationFacts) AttemptICEProfile {
	endpoints := make([]Endpoint, 0, len(profile.endpoints))
	for _, endpoint := range profile.endpoints {
		if endpointEligible(endpoint, request, facts, nil, nil) {
			endpoints = append(endpoints, endpoint)
		}
	}
	return profileFromEndpoints(endpoints, request)
}

func profileFromEndpoints(endpoints []Endpoint, request SelectionRequest) AttemptICEProfile {
	ids := make([]string, len(endpoints))
	for i, endpoint := range endpoints {
		ids[i] = endpoint.ID
	}
	seed := request.NetworkGenerationID + "|" + request.WaveID
	return AttemptICEProfile{id: fmt.Sprintf("ice-%08x", policyHash(seed+"|"+strconv.FormatUint(request.Sequence, 10)+"|"+strings.Join(ids, ","))), generation: request.NetworkGenerationID, endpoints: endpoints}
}

func SelectAttemptProfile(pool ICEEndpointPool, request SelectionRequest, facts GenerationFacts) AttemptICEProfile {
	used := make(map[string]bool)
	domains := make(map[string]bool)
	for _, id := range request.UsedEndpointIDs {
		used[id] = true
	}
	for _, domain := range request.UsedFailureDomains {
		if domain != "" {
			domains[domain] = true
		}
	}
	remaining := min(EndpointsPerProfile, max(0, EndpointsPerWave-len(used)))
	candidates := make([]Endpoint, 0, len(pool.endpoints))
	for _, endpoint := range pool.endpoints {
		if endpointEligible(endpoint, request, facts, used, domains) {
			candidates = append(candidates, endpoint)
		}
	}
	seed := request.NetworkGenerationID + "|" + request.WaveID
	sort.Slice(candidates, func(i, j int) bool { return endpointLess(candidates[i], candidates[j], seed) })
	selected := make([]Endpoint, 0, remaining)
	for len(candidates) > 0 && len(selected) < remaining {
		index := diverseEndpointIndex(candidates, selected)
		selected = append(selected, candidates[index])
		candidates = append(candidates[:index], candidates[index+1:]...)
	}
	return profileFromEndpoints(selected, request)
}

func endpointEligible(endpoint Endpoint, request SelectionRequest, facts GenerationFacts, used, domains map[string]bool) bool {
	if !endpoint.Enabled || used[endpoint.ID] || (endpoint.Family == "ipv4" && !request.IPv4) || (endpoint.Family == "ipv6" && !request.IPv6) || (!request.IPv4 && !request.IPv6) {
		return false
	}
	// A backup explores a new failure domain when that fact is known.
	if endpoint.FailureDomain != "" && domains[endpoint.FailureDomain] {
		return false
	}
	if facts.NetworkGenerationID != request.NetworkGenerationID {
		return true
	}
	for _, fact := range facts.Endpoints {
		if fact.EndpointID == endpoint.ID && request.Now.Before(fact.FailedUntil) {
			return false
		}
	}
	return true
}
func endpointLess(a, b Endpoint, seed string) bool {
	if a.Trust != b.Trust {
		return a.Trust == "reviewed"
	}
	if a.Priority != b.Priority {
		return a.Priority > b.Priority
	}
	ah, bh := policyHash(seed+"|"+a.ID), policyHash(seed+"|"+b.ID)
	if ah != bh {
		return ah < bh
	}
	return a.ID < b.ID
}
func diverseEndpointIndex(candidates, selected []Endpoint) int {
	if len(selected) == 0 {
		return 0
	}
	index, best := 0, -1
	for i, endpoint := range candidates {
		score := 0
		if endpoint.FailureDomain != "" && endpoint.FailureDomain != selected[0].FailureDomain {
			score += 2
		}
		if endpoint.Region != "" && endpoint.Region != selected[0].Region {
			score++
		}
		if score > best {
			index, best = i, score
		}
	}
	return index
}
func policyHash(value string) uint32 {
	h := fnv.New32a()
	_, _ = h.Write([]byte(value))
	return h.Sum32()
}

File: web/src/connectivity/ice-policy/endpoints.ts
export const ENDPOINTS_PER_PROFILE = 2
export const ENDPOINTS_PER_WAVE = 4
export const MAXIMUM_CATALOG_ENDPOINTS = 512
export const ENDPOINT_FAILURE_COOLDOWN_MS = 30_000

export interface ICEEndpoint {
 readonly id: string
 readonly url: string
 readonly region: string
 readonly failureDomain: string
 readonly provider: string
 readonly family: 'ipv4' | 'ipv6' | 'any'
 readonly trust: 'reviewed' | 'local' | 'unreviewed'
 readonly priority: number
 readonly enabled: boolean
}

// The caller is the trusted build/deployment/local configuration boundary.
// Never construct this directory from signaling or capability link data.
export class ICEEndpointPool {
 readonly endpoints: readonly ICEEndpoint[]
 constructor(entries: readonly ICEEndpoint[] = []) {
  const seen = new Set<string>()
  if (entries.length > MAXIMUM_CATALOG_ENDPOINTS) throw new Error('invalid trusted ICE endpoint configuration')
  this.endpoints = Object.freeze(entries.map(endpoint => {
   const match = /^stun:(\[[^\]]+\]|[^:/?#@\s]+):(\d+)$/.exec(endpoint.url)
   const port = Number(match?.[2])
   if (!endpoint.id || seen.has(endpoint.id) || !match || port < 1 || port > 65535 ||
    !['ipv4', 'ipv6', 'any'].includes(endpoint.family) ||
    !['reviewed', 'local', 'unreviewed'].includes(endpoint.trust) ||
    (endpoint.enabled && endpoint.trust === 'unreviewed') || !Number.isInteger(endpoint.priority) || endpoint.priority < 0) {
    throw new Error('invalid trusted ICE endpoint configuration')
   }
   seen.add(endpoint.id)
   return Object.freeze({ ...endpoint })
  }))
 }
}

export interface SelectionRequest {
 readonly networkGenerationID: string
 readonly waveID: string
 readonly sequence: number
 readonly nowMs: number
 readonly usedEndpointIDs: readonly string[]
 readonly usedFailureDomains?: readonly string[]
 readonly ipv4: boolean
 readonly ipv6: boolean
}
export interface GenerationFacts {
 readonly networkGenerationID: string
 readonly endpoints: readonly { readonly endpointID: string; readonly failedUntilMs: number }[]
 readonly profiles?: readonly { readonly profileID: string; readonly serverReflexiveProduced: boolean; readonly firstCandidateDelayMs: number }[]
}
export interface AttemptICEProfile {
 readonly id: string
 readonly networkGenerationID: string
 readonly endpointIDs: readonly string[]
 readonly failureDomains: readonly string[]
 readonly urls: readonly string[]
}
export function policyHash(value: string): number {
 let hash = 0x811c9dc5
 for (const byte of new TextEncoder().encode(value)) hash = Math.imul(hash ^ byte, 0x01000193)
 return hash >>> 0
}
export function selectAttemptProfile(pool: ICEEndpointPool, request: SelectionRequest, facts?: GenerationFacts): AttemptICEProfile {
 const used = new Set(request.usedEndpointIDs)
 const domains = new Set((request.usedFailureDomains ?? []).filter(Boolean))
 const remaining = Math.min(ENDPOINTS_PER_PROFILE, Math.max(0, ENDPOINTS_PER_WAVE - used.size))
 const seed = request.networkGenerationID + '|' + request.waveID
 const candidates = pool.endpoints.filter(endpoint => endpoint.enabled && !used.has(endpoint.id) &&
  (request.ipv4 || request.ipv6) && (endpoint.family !== 'ipv4' || request.ipv4) &&
  (endpoint.family !== 'ipv6' || request.ipv6) && (!endpoint.failureDomain || !domains.has(endpoint.failureDomain)) &&
  !(facts?.networkGenerationID === request.networkGenerationID && facts.endpoints.some(fact => fact.endpointID === endpoint.id && request.nowMs < fact.failedUntilMs)))
 candidates.sort((a, b) => {
  if (a.trust !== b.trust) return a.trust === 'reviewed' ? -1 : 1
  if (a.priority !== b.priority) return b.priority - a.priority
  return policyHash(seed + '|' + a.id) - policyHash(seed + '|' + b.id) || (a.id < b.id ? -1 : 1)
 })
 const selected: ICEEndpoint[] = []
 while (candidates.length && selected.length < remaining) {
  selected.push(candidates.splice(diverseEndpointIndex(candidates, selected[0]), 1)[0]!)
 }
 const endpointIDs = Object.freeze(selected.map(endpoint => endpoint.id))
 return Object.freeze({
  id: 'ice-' + policyHash(seed + '|' + request.sequence + '|' + endpointIDs.join(',')).toString(16).padStart(8, '0'),
  networkGenerationID: request.networkGenerationID,
  endpointIDs,
  failureDomains: Object.freeze(selected.map(endpoint => endpoint.failureDomain)),
  urls: Object.freeze(selected.map(endpoint => endpoint.url)),
 })
}

function diverseEndpointIndex(candidates: readonly ICEEndpoint[], first: ICEEndpoint | undefined): number {
 if (!first) return 0
 let index = 0
 let best = -1
 for (let i = 0; i < candidates.length; i++) {
  const endpoint = candidates[i]!
  const score = (endpoint.failureDomain && endpoint.failureDomain !== first.failureDomain ? 2 : 0) +
   (endpoint.region && endpoint.region !== first.region ? 1 : 0)
  if (score > best) { index = i; best = score }
 }
 return index
}
